### PR TITLE
GEOS-11461: Support MapML embedded viewer on HTML for WFS GetFeature

### PR DESCRIPTION
**Pull request by dromagnoli**
[GEOS-11461](https://osgeo-org.atlassian.net/browse/GEOS-11461)

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### Notes:
The exisiting MapML HTML Embedded Viewer for WMS will thrown an invalid SRS exception if the requested SRS is not one of the 4 supported by MapML. The same logic has been applied when building the MapML HTML Embedded viewer for WFS.

WFS 1.0.0 requires a simple format name to be reported in the capabilities document (i.e. MAPML-HTML). This will reported in 1.1.0 and 2.0.0 too, together with the **text/html; subtype=mapml**

### CODE REVIEW

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLGetFeatureOutputFormat.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    private ProjType parseProjType(Request request) throws ServiceException {
+        try {
+            Map<String, Object> rawKvp = request.getRawKvp();
+            String srs;
+            if (rawKvp.containsKey("SRSNAME")) {
+                srs = (String) rawKvp.get("SRSNAME");
+            } else if (rawKvp.containsKey("SRS")) {
+                srs = (String) rawKvp.get("SRS");
+            } else {
+                srs = "EPSG:4326";
```
Comment (dromagnoli) [inline src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLGetFeatureOutputFormat.java:128]: maybe we should default to OSMTILE?

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLConstants.java [hunk 2]
```diff
@@ -31,6 +33,9 @@ public final class MapMLConstants {
     /** format name */
     public static final String FORMAT_NAME = "MAPML";
 
+    /** format name needed to have a parseable format in WFS 1.0.0 capabilities */
+    public static final String HTML_FORMAT_NAME = "MAPML-HTML";
+
```
Comment (dromagnoli) [inline src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLConstants.java:38]: WFS 1.0.0 requires a simple format name to be reported in the capabilities document. 
This will reported in 1.1.0 and 2.0.0 too, together with the text/html; subtype=mapml

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLGetFeatureOutputFormat.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+            // string value for the srs/crs parameter
+            String parameterName =
+                    Optional.ofNullable(request.getVersion())
+                            .filter(v -> v.equals("1.0.0"))
+                            .map(v -> "srs")
+                            .orElse("srsName");
+            Map<String, Object> rawKvp = Dispatcher.REQUEST.get().getRawKvp();
+            String value = (String) rawKvp.get("SRSNAME");
+            if (value == null) value = (String) rawKvp.get("SRS");
+            throw new ServiceException(
```
Comment (dromagnoli) [inline src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLGetFeatureOutputFormat.java:142]: Here, we are using the same logic that is applied by WMS when producing the HTML. Only the 4 MapML supported CRSs  will be accepted.

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLGetFeatureOutputFormat.java (hunk 1 above)
Comment (prushforth) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLGetFeatureOutputFormat.java:128]: I gather that because OSMTILE does not include the poles, that Andrea prefers to use WGS84.  I can live with that. On the other hand, "OSMTILE" is the default map projection used by the \<mapml-viewer> if you don't specify the `projection` attribute.  That shouldn't impact GeoServer though, since as Andrea points out the OSMTILE can't cover all earthly data.  
`WGS84` could be the default here.

Comment (prushforth) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLGetFeatureOutputFormat.java:128]: Actual live mapml should be served as `text/mapml`.  The preview application is (now, in WMS) served as `text/html; subtype=mapml`.

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLGetFeatureOutputFormat.java (hunk 3 above)
Comment (prushforth) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLGetFeatureOutputFormat.java:142]: Without having time to get into the code, I must point out that we currently can serve any WFS SRS into MapML, realizing that we can't view that content without a custom TCRS definition, but it still may be useful to serve it for completeness' sake.  The technique I've used in the viewer is to disallow / throw on `projection` values that contain a `:` colon.  So when serving a text/mapml document as EPSG:1234 for example, we could create a custom TCRS definition for `EPSG1234` and serve that.  If the projection is unknown i.e. no custom TCRS definition exists, the viewer reports a different problem. The purpose of this is to avoid having people think that EPSG:codes code be uses as TCRS names directly.  i.e. you need the extra defintional stuff provided by a TCRS, it's not the same thing as an EPSG:code.

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLConstants.java (hunk 2 above)
Comment (prushforth) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLConstants.java:38]: what about text/html; subtype=mapml?  This was used in the WMS preview for example.  The important thing is to serve content (when asked for mapml) as `text/mapml`

Comment (dromagnoli) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLConstants.java:38]: Correct. We are indeed using "text/html; subtype=mapml" to serve WFS as HTML (doing the same thing that WMS does).
The problem is in WFS 1.0.0 that is not supporting that as element name for the capabilities due to this check.

https://github.com/geoserver/geoserver/blob/main/src/wfs/src/main/java/org/geoserver/wfs/WFSGetFeatureOutputFormat.java#L125

So we ended up exposing it as MAPML-HTML.
image

The WFS KML output format does a similar thing, using KML for WFS 1.0.0 (see above screenshot) 
This will result in having variations of the same output listed in WFS 1.1.0 and 2.0.0.

Check the below 1.1.0 WFS capabilities document where KML output is listed with the 3 variations (marked in red) and MapML-HTML in 2 variations (marked in yellow)
image

Long story short:
When requesting WFS for `MAPML`, it will indeed serve content as `text/mapml`
When requesting WFS 1.0.0 for `MAPML-HTML` or `text/html; subtype=mapml` it will indeed serve the Mapml client as `text/html; subtype=mapml`.

Does it clarify?

Comment (prushforth) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLConstants.java:38]: Yes it does, thank you very much!

Comment (dromagnoli) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLConstants.java:38]: > Yes it does, thank you very much! So the MAPML value gets sent under the `text/mapml` Content-type: text/mapml. Curious why text/csv didn't get caught by that rule.

Actually it gets caught too.
WFS 1.0.0 only reports `CSV `whilst WFS 1.1.0 reports both `csv `and `text/csv`

image

**Comment by dromagnoli** [comment]
Hi @prushforth , based on your feedbacks, I will create the official PR in GeoServer if no changes are required.
About the CoordinateReferenceSystems management and supported CRS, we will revisit that part once we will work on the other task related to supporting TCRS configurations. 
Please, let us know.

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLConstants.java (hunk 2 above)
Comment (prushforth) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLConstants.java:38]: OK I understand now but in no case do we see `text/mapml` in the list.

Comment (aaime) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLConstants.java:38]: Indeed you don't, it's already the case in an un-modified GeoServer, e.g.:

https://gs-main.geosolutionsgroup.com/geoserver/ows?service=WFS&version=1.1.0&request=GetCapabilities

**Comment by prushforth** [comment]
> Please, let us know

I apologize I have been looking at Joseph's PR, I will just build this one before I give the thumb's up :-)

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLGetFeatureOutputFormat.java (hunk 1 above)
Comment (prushforth) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLGetFeatureOutputFormat.java:128]: > Should we default to OSMTILE?

I should be more clear: I think we should default to WGS84 to be consistent with how it's done on the WMS side.

**Comment by prushforth** [comment]
> About the CoordinateReferenceSystems management and supported CRS, we will revisit that part once we will work on the other task related to supporting TCRS configurations.

Sounds good.  Apart from the defaulting to OSMTILE, I am quite happy with this!

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLGetFeatureOutputFormat.java (hunk 1 above)
Comment (dromagnoli) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLGetFeatureOutputFormat.java:128]: Correct, we are defaulting to WGS84 indeed.

**Comment by dromagnoli** [comment]
> > About the CoordinateReferenceSystems management and supported CRS, we will revisit that part once we will work on the other task related to supporting TCRS configurations.
> 
> Sounds good. Apart from the defaulting to OSMTILE, I am quite happy with this!

Ok. Note that we weren't proposing to default to OSMTILE. That was a question since we were defaulting to WGS84.

**Comment by prushforth** [comment]
Sorry its confusion on my part due to getting up at 3:30 am this week

please proceed, lgtm

**Comment by aaime** [comment]
Dang, merged by mistake. Will clean up the history of the geosolutions fork.
